lib/puppet/provider/foreman_subnet/rest.rb

```diff
@@ -0,0 +1,263 @@
+Puppet::Type.type(:foreman_subnet).provide(:rest) do
+
+  confine :feature => :foreman_api
+
+  def subnets
+    ForemanApi::Resources::Subnet.new({
+      :base_url => resource[:base_url],
+      :oauth    => {
+        :consumer_key    => resource[:consumer_key],
+        :consumer_secret => resource[:consumer_secret]
+      }
+    }, {
+      :headers => {
+        :foreman_user => resource[:effective_user]
+      }
+    })
+  end
+
+  def subnet
+    if @subnet
+      @subnet
+    else
+      spi = subnets.index[0]
+      if spi.is_a?(Hash) && spi.has_key?('results')
+        # foreman_api 0.1.18+
+        @subnet = spi['results'].find { |s| s['name'] == resource[:name] }
+      else
+        # foreman_api < 0.1.18
+        @subnet = spi.find { |s| s['subnet']['name'] == resource[:name] }['subnet']
+      end
+    end
+  end
+
+  def id
+    subnet ? subnet['id'] : nil
+  end
+
+  def exists?
+    !!subnet
+  end
+
+  def create
+    subnets.create({
+      'subnet' => {
+        'name' => resource[:name],
+        'network' => resource[:network],
+        'mask' => resource[:netmask],
+        'gateway' => resource[:gateway],
+        'dns_primary' => resource[:dns_primary],
+        'dns_secondary' => resource[:dns_secondary],
+        'from' => resource[:from],
+        'to' => resource[:to],
+        'vlanid' => resource[:vlan],
+        'dns_id' => smartproxy_id(resource[:dns_proxy]),
+        'dhcp_id' => smartproxy_id(resource[:dhcp_proxy]),
+        'tftp_id' => smartproxy_id(resource[:tftp_proxy]),
+      }
+    })
+  end
+
+  #def flush
+  #  subnets.update({
+  #    'id' => @property_hash[:id],
+  #    'subnet' => {
+  #      'name' => @property_hash[:name],
+  #      'network' => @property_hash[:network],
+  #      'mask' => @property_hash[:netmask],
+  #      'gateway' => @property_hash[:gateway],
+  #      'dns_primary' => @property_hash[:dns_primary],
+  #      'dns_secondary' => @property_hash[:dns_secondary],
+  #      'from' => @property_hash[:from],
+  #      'to' => @property_hash[:to],
+  #      'vlanid' => @property_hash[:vlan],
+  #      'dns_id' => @property_hash[:dns_proxy],
+  #      'dhcp_id' => @property_hash[:dhcp_proxy],
+  #      'tftp_id' => @property_hash[:tftp_proxy],
+  #    }
+  #  })
+  #end
+
+  def destroy
+    subnets.destroy({'id' => id})
+    @subnet = nil
+  end
+
+  def network
+    subnet ? subnet['network'] : nil
```

Comment:
doesn't make any difference, but `subnet['network'] if subnet` would achieve the same